### PR TITLE
refactor(ui): migrate websocket sagas to typescript

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Migrate websocket sagas to typescript.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Run the UI and make sure you can still fetch/update data over the websocket.

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -52,6 +52,7 @@
     "redux-devtools-extension": "2.13.9",
     "redux-saga": "1.1.3",
     "single-spa-react": "4.1.1",
+    "typed-redux-saga": "1.3.1",
```
Comment (huwshimi) [inline ui/package.json:55]: This provides much better typescript support to sagas.

File: ui/package.json [hunk 2]
```diff
@@ -93,6 +94,7 @@
     "@typescript-eslint/parser": "4.24.0",
     "@wojtekmaj/enzyme-adapter-react-17": "0.6.2",
     "babel-eslint": "10.1.0",
+    "babel-plugin-macros": "3.1.0",
```
Comment (huwshimi) [inline ui/package.json:97]: This is used by `typed-redux-saga` when building the js.

File: ui/src/app/base/hooks.ts [hunk 3]
```diff
@@ -17,6 +17,12 @@ import { actions as messageActions } from "app/store/message";
 import type { RootState } from "app/store/root/types";
 import { kebabToCamelCase } from "app/utils";
 
+declare global {
+  interface Window {
+    ga: (...args: unknown[]) => void;
```
Comment (huwshimi) [inline ui/src/app/base/hooks.ts:22]: Not sure why this randomly started complaining at me (it uses `window.ga` in one of the hooks in this file).

File: ui/src/app/base/sagas/websockets.ts [hunk 4]
```diff
@@ -0,0 +1,627 @@
+import type {
```
Comment (huwshimi) [inline ui/src/app/base/sagas/websockets.ts:1]: Not sure why this is showing up as a new file instead of a rename. It might help to manually diff this against the old file.

File: ui/tsconfig.json [hunk 5]
```diff
@@ -19,7 +19,8 @@
     "resolveJsonModule": true,
     "isolatedModules": true,
     "noEmit": true,
-    "jsx": "react-jsx"
+    "jsx": "react-jsx",
+    "types": ["node", "jest"]
```
Comment (huwshimi) [inline ui/tsconfig.json:23]: This is a drive-by to show jest types in .test.ts files.

File: ui/src/app/base/sagas/websockets.ts [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+ * Handle incoming notify messages.
+ *
+ * Notify messages have an action and a payload:
+ * {"type": 2,
+ *  "name": "config",
+ *  "action": "update",
+ *  "data": {"name": "maas_name", "value": "maas-hysteria"}}
+ *
+ * Although we receive a corresponding response for each websocket requests,
+ * the store is only updated once a notify message has been received.
```
Comment (Caleb-Ellis) [inline ui/src/app/base/sagas/websockets.ts:207]: This isn't always true. Maybe "in some cases the store is only..."

File: ui/src/websocket-client.ts [hunk 7]
```diff
@@ -14,17 +12,80 @@ export enum WebSocketMessageType {
...
+    batch?: boolean;
+    // Whether the request should only be fetched the first time.
+    cache?: boolean;
+    // A key to be used to identify a file in the file context.
+    fileContextKey?: string;
+    // Whether the response contains JSON that needs to be parsed.
+    jsonResponse?: boolean;
+    // The endpoint method e.g. "list".
+    method: string;
+    // The endpoing model e.g. "machine".
```
Comment (Caleb-Ellis) [inline ui/src/websocket-client.ts:73]:
```suggestion
    // The endpoint model e.g. "machine".
```

File: ui/src/websocket-client.ts [hunk 8]
```diff
@@ -14,17 +12,80 @@ export enum WebSocketMessageType {
...
+    model: string;
+    // Whether the request should only be fetched the every time.
```
Comment (Caleb-Ellis) [inline ui/src/websocket-client.ts:75]:
```suggestion
    // Whether the request should be fetched every time.
```

**Review by Caleb-Ellis — APPROVED** [review]
LGTM! I think it's a great to have all these extra comments since this section of the codebase is pretty inscrutable.

File: ui/src/app/base/sagas/websockets.ts (hunk 6 above)
Comment (huwshimi) [reply ui/src/app/base/sagas/websockets.ts:207]: Done.

File: ui/src/websocket-client.ts (hunk 7 above)
Comment (huwshimi) [reply ui/src/websocket-client.ts:73]: I can spell!

File: ui/src/websocket-client.ts (hunk 8 above)
Comment (huwshimi) [reply ui/src/websocket-client.ts:75]: And I can grammar too it appears!
